### ISSUE TITLE
Import > BrickLink Studio File... feature

### ISSUE THREAD

**Issue by paramecie** [rgriebl/brickstore#240]
Yes, you can do it :-)

This is the contents of a Studio file I just opened:
![image](https://user-images.githubusercontent.com/32078594/108630447-371b7f80-7465-11eb-8b3c-822efa6a7966.png)

It's a zip file, password protected.
The password, I know it :)
It was given on line during its creation by one of the developpers if I remember well.

It has never changed for 5 years, I just tried it again a few minutes ago :)

So, importing a studio file is simple:
* unzip
* import the proper ldr file

Tada!

**Comment by paramecie** [comment]
Niiice!
As Studio has become the modern LDD, I'm sure it'll help thousands of builders...

2 remarks:

If the studio file is "big", it takes a minute or + to import (50K parts, Core i7) with only a rotating busy cursor showing.
BS seems frozen, but it isn't in fact.
Not much you can do however apart maybe getting the included LDraw file size and at a certain size display a message "This may take some time, please be patient..."?
BTW, that could be the case with other imports, like LDCad...
Just a remark.

It shows nice chevrons for unknown parts, which is fine!
Just to let you know, in case you didn't test it with unknown parts:
![image](https://user-images.githubusercontent.com/32078594/109427090-0f8e5f00-79f1-11eb-9f19-39bc3746872a.png)

All this is very good, thanks!

**Comment by rgriebl** [comment]
I tested it with my Actros (https://brickforge.de/mocs/actros.en.html) which has 4K parts. Import is nearly instant on my i7.
The problem that I'm seeing is Studio itself: loading and saving this model takes ages and editing is very choppy. LeoCAD is waaaaay smoother. I'm wondering how you are able to work with 50K parts in Studio?

**Comment by rgriebl** [comment]
Just a tested a few files from the Studio gallery - I can't see or measure slow loading. Please share a link (or the model itself) ... I'd like to run this through a performance analyzer.

**Comment by ZZJHONS** [comment]
> I tested it with my Actros (https://brickforge.de/mocs/actros.en.html) which has 4K parts. Import is nearly instant on my i7.
> The problem that I'm seeing is Studio itself: loading and saving this model takes ages and editing is very choppy. LeoCAD is waaaaay smoother. I'm wondering how you are able to work with 50K parts in Studio?

What is recommended is to have large builds in different files, and if you want to put it all together, open a new file with the snap option deactivated and then copy all the files content into this one and put together, it should be smoother

**Comment by paramecie** [comment]
> What is recommended is to have large builds in different files

I know! It's not mine, it's a client's one with a big model...

For now I really can't disclose this file.

I'll make some tests, like save it AS to see if it produces a more compact file maybe.

I just tested it again, maybe it was exaggerated when I posted or something has changed; but it takes 30 seconds now to import. Not exactly, I've to make a serious benchmark and more testing...

Anyway I'll have some other trials, I'll let you know @rgriebl ...

**Comment by rgriebl** [comment]
Try build 322. It has some optimization in the ldraw parser. 
It will also spit out some timings in the error log, likes this:

WARN default: unpack/decrypt studio zip: 0'000.000
WARN default: parse ldraw model: 0'000.000
WARN default: parse sub-dat: 0'000.000
...

These numbers are the actual **micro** seconds for loading my 4K parts Actros on Linux ... that's why I'm so surprised that it is so slow for you.

This one here https://www.bricklink.com/v3/studio/design.page?idModel=92730 has 25K parts and loads in 2 micro-seconds for me.

**Comment by paramecie** [comment]
I tried Build 322.
I can't see the WARNs you show above.

BS freezes for 1+ minute, then suddenly I've a bunch of those, and it warns me 23 lots can't be imported (but this is known for me and 'normal'):

![image](https://user-images.githubusercontent.com/32078594/109749955-db2ac680-7bdb-11eb-9236-a0240bef3eec.png)

During the import duration, BS uses maybe 13% of its CPU time, which is less than expected (not 100% at all).

Tomorrow I'll do more tests, including opening the model 92730 above.

More precise info, as I didn't measure anything before.

Our 50K parts studio file:
* studio takes 40 s to open and first render completely this file...
* Leocad opens the first LDR file inside the studio file in maybe 3 seconds (I love Leocad!)
* BS takes 1'50 s (almost 2 minutes) to ask me if I want to import it even with the 23 errors...


Your ANZ studio just downloaded:
* studio opens it in 12 s
* Leocad opens and render in half a second maybe?
* BS : I can't measure, it goes too fast:
WARN default: unpack/decrypt studio zip: 0'000.000
WARN default: parse ldraw model: 0'000.000
WARN default: parse sub-dat: 0'000.000

Not 100% sure at what I can think, especially at this hour...

**Comment by paramecie** [comment]
Weird weird.

Downloaded this vessel, 58K parts (an LDraw file):
https://www.rockraidersunited.com/files/digital_models/ldraw/11-lms-explorer-r195/

Imported in BS: zero second:
WARN default: parse ldraw model: 0'000.000
WARN default: parse sub-dat: 0'000.000

Import and first render in studio: 45 s
Leocad, maybe 3 s

It seems there's something different in my first big file.
Its open time is coherent in Studio and Leocad, rougly proportional as the time of this vessel with 58K parts.
But for BS the import time is VERY abnormal.

There must be some item/combination it doesn't like? I'll try to find out...

**Comment by rgriebl** [comment]
> BS freezes for 1+ minute, then suddenly I've a bunch of those, and it warns me 23 lots can't be imported (but this is known for me and 'normal'):
> 
> ![image](https://user-images.githubusercontent.com/32078594/109749955-db2ac680-7bdb-11eb-9236-a0240bef3eec.png)
> 
> During the import duration, BS uses maybe 13% of its CPU time, which is less than expected (not 100% at all).

Can you copy&paste  me the top most debug output ... the 2 lines with `unpack/decrypt studio zip` and `parse ldraw model`?
(or just even all of them -- you can copy with right-click->copy in the error log)

**Comment by paramecie** [comment]
> Can you copy&paste me the top most debug output ..

No, they're not here.
I guess the log window has a clean up mecanism, there are too many WARN parse sub-dat, it's all I can see.
Temporary release this limit?

**Comment by paramecie** [comment]
I made some tests.
I removed more than the half of the parts (remains maybe 15K...)
Still taking 30 s to load on BS :-(

I unzipped the studio file in a folder.
I tried to import the model.ldr, same thing :-(

I noticed many references to submodels like "grass.ldr" (more than 500).

I replaced all the .ldr by 3001.ldr (so it's now a part and not a submodel).

It made a huge change, and could be loaded very quickly:

WARN default: parse ldraw model: 0'000.000
WARN default: parse sub-dat: 0'000.000
WARN default: parse sub-dat: 0'000.000
WARN default: parse sub-dat: 0'000.000
... (167 warnings)

What I guess is that the parsing is making a back and forth for every submodel appearance (like calling a function), and if there are many submodels (with some having submodels inside...), maybe some with errors (placement or color), then the import time explodes...

In short, a 50K parts model with 1 or 2 submodels opens instantly, but with 1000 it'll take a few minutes...

Also, it doesn't seem to be a problem for studio or leocad; they open this file with the same delay having 1 or 1000 submodels. That's maybe related to the lib you're using for parsing?

Just thoughts...

**Comment by paramecie** [comment]
That could confirm, I looked at the ANZ file and the 58K parts vessel above.
They both don't have any submodel.
They import instantly.

**Comment by paramecie** [comment]
I made a studio test file, see attachment.

It's made of:
* one small submodel of 2 plates
* one submodel of 1000 of those small submodel of 2 plates
* one main model of 10 of the 1000x2

![image](https://user-images.githubusercontent.com/32078594/109812289-2e2c6a00-7c2c-11eb-810c-0835a4c5d88b.png)

It's a pain to open on Studio, it takes 1 minute to open, almost no response, and a minute to close...

Leocad opens and render it *instantly*.
I don't know how Leocad does this, but apparently it has a smart algorithm...

BrickStore imports it in 3 s approx, but throws a bunch of WARN default: parse sub-dat: 0'000.000

That doesn't explain the 2 minutes load for the other file, which may be related to the number of errors encountered in numerous submodels... maybe like 1000 x 0.1 s  = almost 2 minutes...

Anyway...

BTW, the author of Leocad is Leonardo Zide, Senior programmer of Call of Duty (Black Ops and such).
https://www.mobygames.com/developer/sheet/view/developerId,147479/
:0)

You've to unzip this first.
[tests submodels.zip](https://github.com/rgriebl/brickstore/files/6076236/tests.submodels.zip)

**Comment by paramecie** [comment]
FYI, build 324, no change.

**Comment by rgriebl** [comment]
The 0 micro seconds sure looked nice, but they weren't the truth: the compiler optimized the timer away ;)

I just added sub-model caching now ... not pushed yet, but here are the numbers (for real this time)

submodels test file from Sylvain (a lot of sub-models)
> WARN default: unpack/decrypt studio zip: 0'001.697
> WARN default: parse ldraw model: 0'003.517
> WARN default: removing duplicates: 0'000.063
> WARN default: consolidate ldraw model: 0'000.060

24K items tower from BL Studio Gallery (no sub-models)
> WARN default: unpack/decrypt studio zip: 0'011.387
> WARN default: parse ldraw model: 0'065.238
> WARN default: removing duplicates: 0'216.144
> WARN default: consolidate ldraw model: 0'015.750

My 4K Actros (a few sub-models)
> WARN default: unpack/decrypt studio zip: 0'003.334
> WARN default: parse ldraw model: 0'422.612
> WARN default: removing duplicates: 0'005.035
> WARN default: consolidate ldraw model: 0'003.456

So the most expensive one is now my Actros with a lot of individual parts, but still at under 0.5 seconds total.

**Comment by paramecie** [comment]
OK Boss - build it and I'll test this :)

**Comment by paramecie** [comment]
Build 325

**submodels test file from me (a lot of sub-models)**
WARN default: unpack/decrypt studio zip: 0'001.369
WARN default: parse ldraw model: 0'003.290
WARN default: removing duplicates: 0'000.081
WARN default: consolidate ldraw model: 0'000.076

**ANZ Tower Lvls 1 through 4_Copy**
WARN default: unpack/decrypt studio zip: 0'012.694
WARN default: parse ldraw model: 0'052.449
WARN default: removing duplicates: 0'267.677
WARN default: consolidate ldraw model: 0'015.966

**1:1 L.M.S. Explorer**
WARN default: parse ldraw model: 0'155.627
WARN default: removing duplicates: 1'546.795
WARN default: consolidate ldraw model: 0'028.893
WARN default: Image download failed: "Error transferring https://img.bricklink.com/ItemImage/PN/10/60479.png - server replied: Not Found" ( 404 )
WARN default: Image download failed: "Error transferring https://img.bricklink.com/ItemImage/PN/10/3026.png - server replied: Not Found" ( 404 )
WARN default: Image download failed: "Error transferring https://img.bricklink.com/ItemImage/PN/10/4460.png - server replied: Not Found" ( 404 )
WARN default: Image download failed: "Error transferring https://img.bricklink.com/ItemImage/PN/10/30144.png - server replied: Not Found" ( 404 )

**Our "2 minutes 50K parts many submodels file"**
WARN default: unpack/decrypt studio zip: 0'012.003
WARN default: parse ldraw model: 40'492.853
WARN default: removing duplicates: 0'714.336
WARN default: consolidate ldraw model: 0'015.518
